core/corehttp/gateway_handler.go

```diff
@@ -324,61 +344,34 @@ func (i *gatewayHandler) getOrHeadHandler(w http.ResponseWriter, r *http.Request
 	logger := log.With("from", r.RequestURI)
 	logger.Debug("http request received")
 
-	// X-Ipfs-Gateway-Prefix was removed (https://github.com/ipfs/go-ipfs/issues/7702)
-	// TODO: remove this after  go-ipfs 0.13 ships
-	if prfx := r.Header.Get("X-Ipfs-Gateway-Prefix"); prfx != "" {
-		err := fmt.Errorf("X-Ipfs-Gateway-Prefix support was removed: https://github.com/ipfs/go-ipfs/issues/7702")
-		webError(w, "unsupported HTTP header", err, http.StatusBadRequest)
+	var (
+		err          error
+		shouldReturn bool
+	)
+
+	if err := handleUnsupportedHeaders(r); err != nil {
+		webRequestError(w, err)
 		return
 	}
 
-	// ?uri query param support for requests produced by web browsers
-	// via navigator.registerProtocolHandler Web API
-	// https://developer.mozilla.org/en-US/docs/Web/API/Navigator/registerProtocolHandler
-	// TLDR: redirect /ipfs/?uri=ipfs%3A%2F%2Fcid%3Fquery%3Dval to /ipfs/cid?query=val
-	if uriParam := r.URL.Query().Get("uri"); uriParam != "" {
-		u, err := url.Parse(uriParam)
-		if err != nil {
-			webError(w, "failed to parse uri query parameter", err, http.StatusBadRequest)
-			return
-		}
-		if u.Scheme != "ipfs" && u.Scheme != "ipns" {
-			webError(w, "uri query parameter scheme must be ipfs or ipns", err, http.StatusBadRequest)
-			return
-		}
-		path := u.Path
-		if u.RawQuery != "" { // preserve query if present
-			path = path + "?" + u.RawQuery
-		}
-
-		redirectURL := gopath.Join("/", u.Scheme, u.Host, path)
-		logger.Debugw("uri param, redirect", "to", redirectURL, "status", http.StatusMovedPermanently)
-		http.Redirect(w, r, redirectURL, http.StatusMovedPermanently)
+	err, shouldReturn = handleProtocolHandlerRedirect(w, r, logger)
+	if err != nil {
+		webRequestError(w, err)
```

Comment:
Why aren't you returning here ? (I already know why, because `handleProtocolHandlerRedirect` cannot return an error without returning `true`)

But that is information, a reader that unfamiliar with this code and the compiler doesn't know.
```suggestion
		webRequestError(w, err)
		return
```
To avoid rejoining the false branch, pls terminate here.

Reply:
Ah, I see.  You're suggesting that errors should always return and then the extra boolean is just used to indicate that non error conditions should return as well.  Gotcha.  That sounds good to me.  👍

Reply:
Yes.

I think that:
```
return error > return bool > return error, bool
```
In that case we can't `return error` because there are non erroring cases that still need a termination, so might as well fully commit to a bool.